traitlets/traitlets.py

```diff
@@ -1362,49 +1369,71 @@ def traits(self, **metadata):
         the output.  If a metadata key doesn't exist, None will be passed
         to the function.
         """
-        traits = dict([memb for memb in getmembers(self.__class__) if
+        traits = dict([memb for memb in getmembers(cls) if
                      isinstance(memb[1], TraitType)])
 
         if len(metadata) == 0:
             return traits
 
+        for meta_name, meta_eval in metadata.items():
+            if type(meta_eval) is not types.FunctionType:
+                metadata[meta_name] = _SimpleTest(meta_eval)
+
         result = {}
         for name, trait in traits.items():
             for meta_name, meta_eval in metadata.items():
-                if type(meta_eval) is not types.FunctionType:
-                    meta_eval = _SimpleTest(meta_eval)
                 if not meta_eval(trait.metadata.get(meta_name, None)):
                     break
             else:
                 result[name] = trait
 
         return result
 
-    def trait_metadata(self, traitname, key, default=None):
+    @classmethod
+    def trait_metadata(cls, traitname, key, default=None):
         """Get metadata values for trait by key."""
         try:
-            trait = getattr(self.__class__, traitname)
+            trait = getattr(cls, traitname)
         except AttributeError:
             raise TraitError("Class %s does not have a trait named %s" %
-                                (self.__class__.__name__, traitname))
+                                (cls.__name__, traitname))
         else:
             return trait.metadata.get(key, default)
 
-    def add_traits(self, **traits):
-        """Dynamically add trait attributes to the HasTraits instance."""
-        self.__class__ = type(self.__class__.__name__, (self.__class__,),
-                              traits)
-        for trait in traits.values():
-            trait.instance_init(self)
+    @classmethod
+    def class_own_events(cls, name):
+        """Get a dict of all event handlers defined on this class, not a parent.
 
-    def set_trait(self, name, value):
-        """Forcibly sets trait attribute, including read-only attributes."""
-        cls = self.__class__
-        if not self.has_trait(name):
-            raise TraitError("Class %s does not have a trait"
-                             "named %s" % (cls.__name__, name))
-        else:
-            getattr(cls, name).set(self, value)
+        Works like ``event_handlers``, except for excluding traits from parents.
+        """
+        sup = super(cls, cls)
+        return {n: e for (n, e) in cls.events(name).items()
+                if getattr(sup, n, None) is not e}
+
+    @classmethod
+    def events(cls, name=None):
```

Comment:
I guess the main reason I didn't do that is because it seemed unnecessary to call it `trait_events` as opposed to just `events`, but that's reasonable if you think that might be too generic.